internal/builtinfunctions/functions.go

```diff
@@ -85,11 +85,20 @@ func getFloatToIntFunction() schema.CallableFunction {
 			schema.PointerTo(
 				"Converts a float type into an integer type by discarding the fraction."+
 					" In other words, it is rounded to the nearest integer towards zero.\n"+
+					"Special cases:\n"+
+					" +Inf outputs the maximum 64-bit integer (9223372036854775807)\n"+
+					" -Inf and NaN output the minimum 64-bit integer (-9223372036854775808)\n\n"+
 					"For example, `5.5` becomes `5`, and `-1.9` becomes `-1`",
 			),
 			nil,
 		),
 		func(a float64) int64 {
+			// Define behavior for special cases. The raw cast in go has platform-specific behavior.
+			if a == math.Inf(1) {
+				return math.MaxInt64
+			} else if a == math.Inf(-1) || math.IsNaN(a) {
+				return math.MinInt64
+			}
```

Comment:
I don't see an advantage either way. But positive seems more valid than negative. And NaN is invalid.

Reply:
Returning an error would in some sense be best, but that requires the workflow to be able to make sense of and handle the error; and in this context it seems unlikely that's feasible. Ah, well: I'll try to stop obsessing about this. In theory I have enough other stuff about which to obsess to keep me busy for a while.

Reply:
There's no way to handle the error within a workflow, but an error is a valid way to do it. It would just make the dependency unresolvable. All I'd need to do is change one valid and return an error return in addition to the integer return.

Reply:
Given that we have defined behaviors for all non-integer-valued inputs to this function, as well as for "infinities" and other weird things (like negative zeros), it seems a shame to me to complicate the interface so that we can fail for the `NaN` case.  I think it's perfectly adequate to make it return `MinInt64` -- it is the least of evils.

Reply:
> We've suffered some quiet math error

More likely, the workload output from the previous step produced one or more NaN's in its output, and the subsequent step is trying to process the data stream.  I don't know why the step is trying to convert float's into int's (that seems like a weird thing to do), but I don't want the entire flow to error-out just because the workload in the previous step polluted the data stream with some NaN's.  (Been there, done that, still carrying the scars....)  If the conversion is allowed to proceed and produce a crazy values, those can easily be detected as outliers and filtered out of the data in a subsequent step (or otherwise addressed)...but, we presumably won't get to that step if the conversions produce errors...so, the NaN's will have to be dealt with further up the pipeline...perhaps that's doable, but it might be less convenient.

> the ARCA expression language doesn't even have the concept currently of a "check for NaN"

I'm not sure what you are thinking of in terms of "checking" -- we don't have an `if` construct, do we?  (And, we don't have a ternary operator, and our AND and OR operators accept and produce only booleans.)  So, even if we could "check" for a NaN, I'm not sure what we could do if we found one.  (That said, I suspect that we can check for NaNs using the IEEE definition of `==` and `!=` both failing for a NaN....)

> I'm uncomfortable with wallpapering over this crack

I think we are filling the crack by providing a well-defined behavior for it.  Admittedly, we are replacing one mode of failure with another, but I'm pretty confident that someone will notice if their data shows up with a value of 9.2 _quadrillion_ in it.

Reply:
> > We've suffered some quiet math error
> 
> More likely, the workload output from the previous step produced one or more NaN's in its output

Um, yes, exactly: because of a math error.

> , and the subsequent step is trying to process the data stream. I don't know why the step is trying to convert float's into int's (that seems like a weird thing to do), but I don't want the entire flow to error-out just because the workload in the previous step polluted the data stream with some NaN's. (Been there, done that, still carrying the scars....) If the conversion is allowed to proceed and produce a crazy values, those can easily be detected as outliers and filtered out of the data in a subsequent step (or otherwise addressed)...but, we presumably won't get to that step if the conversions produce errors...so, the NaN's will have to be dealt with further up the pipeline...perhaps that's doable, but it might be less convenient.

So you'd rather just lie with a whopping big negative number and hope that someone catches it? Huh. I can only hope nobody ever actually generates a NaN in an ARCA workflow, because I suspect someone's going to be seriously surprised and upset.

Or I guess maybe just hope that nobody ever converts floating computation results to int; which, as you say, sounds like a relatively odd thing to do.

Maybe we just shouldn't support float to int. 😀 

> > the ARCA expression language doesn't even have the concept currently of a "check for NaN"
> 
> I'm not sure what you are thinking of in terms of "checking" -- we don't have an `if` construct, do we? (And, we don't have a ternary operator, and our AND and OR operators accept and produce only booleans.) So, even if we could "check" for a NaN, I'm not sure what we could do if we found one. (That said, I suspect that we can check for NaNs using the IEEE definition of `==` and `!=` both failing for a NaN....)

Yeah, I suppose that you could pass a `!expr x == x && x != x` as input to a plugin step along with the actual value and it could normalize according to the workflow's needs ... or fail out. So the user isn't entirely helpless. Still, I think that substituting a bizarre arbitrary integer is a bad thing for the engine to do transparently.

> > I'm uncomfortable with wallpapering over this crack
> 
> I think we are filling the crack by providing a well-defined behavior for it. Admittedly, we are replacing one mode of failure with another, but I'm pretty confident that someone will notice if their data shows up with a value of 9.2 _quadrillion_ in it.

Yeah, maybe someone will notice (depending on how the algorithm uses this integer); but at absolute best, the *cause* of the bizarre symptom will not be remotely obvious at that level.

Reply:
> > > We've suffered some quiet math error
> > 
> > 
> > More likely, the workload output from the previous step produced one or more NaN's in its output
> 
> Um, yes, exactly: because of a math error.

Um, no, it's because the workload (or its post-processing) was obliged to supply a value for which it had no data, so it produced a `"NaN"` because "N/A", "nil", "null", "None", and the like don't readily convert to floating point values.  😝 

> So you'd rather just lie with a whopping big negative number and hope that someone catches it? Huh. I can only hope nobody ever actually generates a NaN in an ARCA workflow, because I suspect someone's going to be seriously surprised and upset.

It's an untenable situation, so we're reduced to choosing the least of evils...and, I think allowing the calculation to proceed (with what might be crazy results) is better than stopping it, because the situation might be able to be rectified downstream.  Arguably, the more correct solution is to avoid converting NaN's to int's, but I don't feel inclined to impose that requirement on our users.

> Or I guess maybe just hope that nobody ever converts floating computation results to int; which, as you say, sounds like a relatively odd thing to do.
> 
> Maybe we just shouldn't support float to int. 😀

It's a thought.  A similar thought is to refuse to convert floating point values which aren't integral -- then we're not only out of the business of dealing with NaNs and infinities, but we also don't have to worry about truncation or rounding.

> Yeah, I suppose that you could pass a `!expr x == x && x != x` as input to a plugin step along with the actual value and it could normalize according to the workflow's needs ... or fail out. So the user isn't entirely helpless. Still, I think that substituting a bizarre arbitrary integer is a bad thing for the engine to do transparently.

I think the proper test is `x != x` -- I believe that that's true for NaN and false for all other values.  Nevertheless, I'm not sure that the step is limited in terms of what it can do to the data values (and, it doesn't need an Arca expression to evaluate them).

> > I think we are filling the crack by providing a well-defined behavior for it. Admittedly, we are replacing one mode of failure with another, but I'm pretty confident that someone will notice if their data shows up with a value of 9.2 _quadrillion_ in it.
> 
> Yeah, maybe someone will notice (depending on how the algorithm uses this integer); but at absolute best, the _cause_ of the bizarre symptom will not be remotely obvious at that level.

I would hope someone would notice (_quadrillion_ is **really** big...)!  But, as you say, the cause might be hard to discern.

I just don't like returning an error in this case...there's no way for the workflow to recover from the error, is there?  So, if there's a NaN in the pipe, the whole thing explodes, taking all the other data with it?  That's disappointing to me.  (Perhaps we _shouldn't_ be providing `floatToInt()`....)

Reply:
> > > > We've suffered some quiet math error
> > > 
> > > 
> > > More likely, the workload output from the previous step produced one or more NaN's in its output
> > 
> > 
> > Um, yes, exactly: because of a math error.
> 
> Um, no, it's because the workload (or its post-processing) was obliged to supply a value for which it had no data, so it produced a `"NaN"` because "N/A", "nil", "null", "None", and the like don't readily convert to floating point values. 😝

That does not seem to me to be a meaningful interpretation of what NaN means. It's generated by IEEE float hardware as a result of an illegal operation like divide-by-zero or square-root-of-negative. The idea is that in complicated expressions the error is carried through non-destructively without losing the fact that the result is garbage. (If you care, you can detect it along the way, but that adds overhead that isn't usually needed.)

Of course, you *can* manually assign NaN (`float('nan')`) as a "missing value", but that's not directly a concern of the standard.

(Of course language libraries often do things like raising `ZeroDivisionError` or `ValueError` (which is explicitly allowed) instead of generating a NaN, but the IEEE standard is dealing with raw hardware, where signalling an error is more disruptive.) But `0.0 * float('inf')` results in `nan` as does `float('inf') / float('inf')`.

> I just don't like returning an error in this case...there's no way for the workflow to recover from the error, is there? So, if there's a NaN in the pipe, the whole thing explodes, taking all the other data with it? That's disappointing to me. (Perhaps we _shouldn't_ be providing `floatToInt()`....)

For whatever anyone might think it worth in this context, IEEE agrees with me: 
"when converting NaN to an integer type, the [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) standard requires that the invalid-operation [exception](https://en.wikipedia.org/wiki/IEEE_754#Exception_handling) be signaled." I just don't see how anything else makes sense, and it risks massive confusion.

With that said, yeah, we're hampered by not having a fluid error detection and handling framework in ARCA; the expression language really isn't designed for computation. But I don't have a better suggestion, and choosing an arbitrary number just seems like almost the worst possible solution.